### PR TITLE
[SYCL] Add template parameter support for no_global_work_offset attribute

### PR DESCRIPTION
**Pull request by smanna12**
This patch

adds support for template parameter on
[[intel:: no_global_work_offset())]] attribute where
valid values are 0 and 1. If 1, compiler doesn't 
use the global work offset values for the device function. 
If used without argument, value of 1 is set implicitly.
Attribute parameter is optional,
so [[intelfpga::no_global_work_offset]] means 
the same as [[intelfpga::no_global_work_offset(1)]].

updates sema/codegen tests with mock headers on device.

uses existing function "sema::addIntelSYCLSingleArgFunctionAttr" from
other single argument function attributes such as num_simd_work_items,
max_global_work_dim, and intel_reqd_sub_group_size to avoid source
codes duplication and reuse the codes for the template parameter support.

Removes Bool argument "Enabled" from test and attr.td file since template 
parameter covers functionality of "Enabled" parameter.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp
Comment (smanna12) [inline clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp:49]: Template parameter support changes sema part, do we still keep "Enabled" here?

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (smanna12) [inline clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:29]: template support does not accept  [[intel::no_global_work_offset]].

Test is updated since 
valid values are 0 and 1 and attribute parameter is optional,
so [[intel::no_global_work_offset]] means the same as
[[intel::no_global_work_offset(1)]].

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (smanna12) [inline clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:18]: template support does not accept [[intel::no_global_work_offset]].

Test is updated since
valid values are 0 and 1 and attribute parameter is optional,
so [[intel::no_global_work_offset]] means the same as
[[intel::no_global_work_offset(1)]].

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (Fznamznon) [inline clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:41]: I think the attribute works in a bit another way now. What was before: if value of "Enabled" parameter was true, then metadata no_global_work_offset was added. Now it looks like it can be added even if attribute argument was "0" but with different number in corresponding metadata. I can't see it in the test, but code looks like it does so. Am I wrong? Anyway, I can clearly see that change - ![[NUM5]] = !{i32 1}, are we supposed change LLVM IR output in this patch?

Comment (smanna12) [reply clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:41]: @Fznamznon thanks for looking into the patch. Yes, you are right that current support breaks the existing codes. Codegen seems working wrong way. Codegen part needs to redesign here with the template support. Any thought?

>>are we supposed change LLVM IR output in this patch?
Probably not.

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (smanna12) [reply clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:18]: Yes, with the current approach,  it is not possible to use the attribute without an argument anymore

Comment (Fznamznon) [reply clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:18]: I think it also can break existing code.

Comment (mlychkov) [reply clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:18]: @GarveyJoe Is it critical that [[intel::no_global_work_offset]] cannot be used without an argument anymore?

File: clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp
Comment (mlychkov) [inline clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp:8]: Is this change intentional?

File: clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp
Comment (mlychkov) [inline clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp:3]:
```suggestion
// Test that checks template parameter support for 'no_global_work_offset' attribute on sycl device.
```

File: clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp
Comment (smanna12) [reply clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp]: No, this is not intentional. Test is changed to original.

**Review by mlychkov — COMMENTED** [review]
LGTM. Thank you.

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (smanna12) [reply clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp]: This is fixed now.

File: clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp
Comment (smanna12) [reply clang/test/SemaSYCL/check-notdirect-attribute-propagation.cpp:49]: Done

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:5308]: Why two separate code paths? I think it'd be more clear to do:
```
Expr *E = Attr.isArgExpr(0) ? Attr.getArgAsExpr(0) : IntegerLiteral::Create(...);
D->addAttr(...);
```

**Review by AaronBallman — COMMENTED** [review]
Some drive-by comments as the review happened to catch my attention.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: @AaronBallman for looking into the patch and feedback. 
Yes, no need to have a separate code paths. I have updated it.

File: clang/include/clang/Sema/Sema.h
Comment (AaronBallman) [inline clang/include/clang/Sema/Sema.h:12976]: Sorry for not noticing this earlier, but I was curious about the warning since it wasn't familiar to me and that caused me to notice a few issues.

https://github.com/intel/llvm/blob/sycl/clang/include/clang/Basic/DiagnosticSemaKinds.td#L11147

0. The diagnostic suggests that the boolean argument here will be adjusted to be `1` automatically, but the code is dropping the attribute entirely.

1. The diagnostic wording doesn't follow the usual style (it starts with a capital letter and contains sensible full stop punctuation). This isn't specific to your patch, but is an inconsistency we may want to do something about (I notice a few other warnings that have similar issues).

2. The warning itself is rather suspect. Generally speaking, a boolean value is either zero or nonzero, so warning "you wrote something nonzero, we've switched it to a different nonzero value" is a bit strange. This also isn't specific to your patch.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:12976]: >>Generally speaking, a boolean value is either zero or nonzero, so warning "you wrote something nonzero, we've switched it to a different nonzero value" is a bit strange. 

Yes, I agree. This diagnostic seems very confusing to me.  This diagnostic was added only for this attribute.

I am inclined to change like this. @AaronBallman what is your thought? This provides clear info that the attribute can only take 0 or one.

if (ArgInt > 1) {
        Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
            << CI.getAttrName() << 0 << 1 << E->getSourceRange();
        return;
      }

Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h:12976]: Backing up a step so I can see the bigger picture -- what is the goal of diagnosing a value other than 0 or 1 here? Is that a requirement from some other standard? Do bad things happen if they pass 2 instead of 1? Or is it more a matter of trying to enforce a sane coding style?

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:12976]: I am not sure about this. Tagging @MrSidims for this comments. He reviewed original attribute implementation.

Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h:12976]: I would imagine that if any nonzero value will do (so it's just a matter of style), we should accept silently and not bother warning. If the value has to be zero or one because of assumptions elsewhere in the compiler, we could force the value 0 or 1 when creating the semantic attribute (so if the user passes in `2`, we manually create an integer literal for `1` instead). If the value has to be zero or one because of portability to other toolchains, requirements from another standard, etc then diagnosing makes more sense to me.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:12976]: Tagging @MrSidims/@GarveyJoe for this comment.

Comment (keryell) [reply clang/include/clang/Sema/Sema.h:12976]: > Backing up a step so I can see the bigger picture -- what is the goal of diagnosing a value other than 0 or 1 here? Is that a requirement from some other standard? Do bad things happen if they pass 2 instead of 1? Or is it more a matter of trying to enforce a sane coding style?

I think you pointed the very problem. The API is unclear. If a boolean is expected, then it should be `[[intelfpga::no_global_work_offset(true)]]` and `[[intelfpga::no_global_work_offset(false)]]`, which removes the need for testing anything less just by relying on usual C/C++ conventions.

Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h:12976]: I don't think this diagnostic carries its weight. It introduces an inconsistency with how other attributes are handled (which makes all attributes harder to use because you have to guess which attributes work what way) and doesn't follow the usual C or C++ conventions (where 0 is false and nonzero is true). Also, while the SYCL attributes currently focus on C++, part of the consistency with C means that `true` and `false` are actually problematic in some C language modes because those aren't keywords (you need to include a header file to use macros with that name, and the macros expand to integers anyway).

I'd recommend removing this diagnostic entirely.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:12976]: @AaronBallman for the recommendation.  I will remove this entirely. 

I have one question, since attribute takes only 0 and 1, if user passes 2 or more as attribute value, should we give the diagnostic like and ignore the attribute. We use this for other attributes also.

if (ArgInt > 1) {
Diag(E->getBeginLoc(), diag::err_attribute_argument_out_of_range)
<< CI.getAttrName() << 0 << 1 << E->getSourceRange();
return;
}

Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h:12976]: I wouldn't diagnose that situation at all because the attribute takes a Boolean expression, and `2` is a valid Boolean expression that means `true`. If it makes your life easier, you could correct the value when creating the semantic attribute. e.g.,
```
Expr *Arg = nullptr;
if (Expr *E = Attr.getArgAsExpr(0)) {
  uint32_t Val;
  if (!checkUint32Argument(S, AL, E, Val))
    return;
  // The optional argument is present but does not evaluate to 0 or 1,
  // so force the value to be 1.
  if (Val && Val != 1)
    Arg = IntegerLiteral::Create(...); // Create literal 1
  else
    Arg = E; // The literal the user wrote is fine, just use it.
} else
  Arg = IntegerLiteral::Create(...); // Create literal 1 for the default value

D->addAttr(..., Arg, ...);
```
Note, I'd only jump through these hoops if you absolutely have to as it makes the code more complicated to reason about. A better approach, if possible, is to handle it when getting the value out of the expression attached to the semantic attribute. At some point you'll be getting an `APInt` object out of the expression and you can check whether the value of that integer is nonzero or not at that point.

Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: I agree. This will make the code more complicated. I have removed the diagnostic entirely.  If there is a reason, we can make the above changes and could correct the value when creating the semantic attribute. 

I see similar HLS attribute that supports GNU syntax and uses old spelling of SYCL attribute (uses_global_work_offset, https://github.com/intel/llvm/pull/1090). This attributes treats the value 2 or so  as "true" Boolean value and generates same AST/IR like value 1.

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp]: Done

File: clang/include/clang/Basic/Attr.td
Comment (smanna12) [reply clang/include/clang/Basic/Attr.td]: Done

File: clang/include/clang/Sema/Sema.h
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: Done

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (AaronBallman) [inline clang/lib/CodeGen/CodeGenFunction.cpp:697]: FWIW, this is checking that the attribute has a non-null `Expr*`, not what the value of the expression is. Should you be checking for non-zero vs zero here?

File: clang/include/clang/Basic/Attr.td
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:1307]: I would expect there to be a `let TemplateDependent = 1;` here to instantiate the attribute within a template. But then again, we're following the template instantiation pattern used by other attributes (which makes me wonder if someday I should go add some more tablegen support for this stuff).

I'd say for right now this is fine as-is, I guess.

File: clang/include/clang/Sema/Sema.h
Comment (AaronBallman) [inline clang/include/clang/Sema/Sema.h:12975]: I don't think this needs to be added here -- we only care about zero vs nonzero for this attribute.

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (AaronBallman) [inline clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:16]: You should also add a test like:
```
template <int N>
[[intel::no_global_work_offset(N)]] void func() {}
```
as both a member function of this class template and as a free function.

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (smanna12) [reply clang/lib/CodeGen/CodeGenFunction.cpp:697]: According to the documentation *https://github.com/smanna12/llvm/blob/83e0f1332cbfcbc73058b11b9c8fb7c260168652/clang/include/clang/Basic/AttrDocs.td#L2334),

If 1, compiler doesn't use the global work offset values for
the device function. Valid values are 0 and 1. If used without argument, value
of 1 is set implicitly. 

I have checked the attribute has non-null EXpr * instead of value of the expression to keep the IR same (clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp) as we had before the template support (PR https://github.com/intel/llvm/pull/1090)

File: clang/include/clang/Basic/Attr.td
Comment (smanna12) [reply clang/include/clang/Basic/Attr.td:1307]: I will leave this now. Inside Attr.td file, i see that we used let TemplateDependent = 1 only for GNU/Clang attribute spellings. Since we are  following the template instantiation pattern used by other SYCL attributes here, i will not add this now.  We can add this later if we decide to go with it.

Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:1307]: Agreed!

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (AaronBallman) [reply clang/lib/CodeGen/CodeGenFunction.cpp:697]: So it sounds like this isn't correct currently (and that suggests there's test coverage that's missing or incorrect) -- `A->getValue()` will always be `true` even if the user wrote `[[intel::no_global_work_offset(0)]]` because it's testing the `Expr*` is not null (and it's always nonnull) rather than the expression's value.

I think the code here should be:
```
const Expr *Arg = A->getValue();
assert(Arg && "Got an unexpected null argument");
llvm::Optional<llvm::APSInt> OptVal = Arg->getIntegerConstantExpr(getContext());
assert(OptVal && "Got something other than an integer constant expression");
if (*OptVal)
  Fn->setMetadata(...);
```

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDeclAttr.cpp:5304]: In the case where `isArgExpr(0)` returns a nonnull pointer, we need to validate that the expression is either dependent (so we'll check it when instantiating the template) or is an integer constant expression.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:780]: And somewhere around here (either within this helper function, if appropriate, or here in `InstantiateAttrs()`) we should be checking that the instantiated expression is an integer constant expression.

File: clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp
Comment (AaronBallman) [inline clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp:45]: Additionally, you should add a test where a template instantiation is just plain wrong:
```
template <typename Ty>
[[intel::no_global_work_offset(Ty{})]] void func() {}

struct S {};
func<S>();
```
as well as a test where the expression is not a constant expression:
```
int foo();
[[intel::no_global_work_offset(foo() + 12)]] void func();
```
and a test where it is a constant expression:
```
constexpr int foo() { return 0; }
[[intel::no_global_work_offset(foo() + 12)]] void func();
```

File: clang/lib/Sema/SemaDeclAttr.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaDeclAttr.cpp:5304]: Inside Sema.h , we have check like this :

template <typename AttrType>
void Sema::addIntelSYCLSingleArgFunctionAttr(Decl *D,
                                             const AttributeCommonInfo &CI,
                                             Expr *E) {
  assert(E && "Attribute must have an argument.");

  if (!E->isInstantiationDependent()) {
    Optional<llvm::APSInt> ArgVal = E->getIntegerConstantExpr(getASTContext());
    if (!ArgVal) {
      Diag(E->getExprLoc(), diag::err_attribute_argument_type)
          << CI.getAttrName() << AANT_ArgumentIntegerConstant
          << E->getSourceRange();
      return;
    }

Comment (AaronBallman) [reply clang/lib/Sema/SemaDeclAttr.cpp:5304]: Ah, thank you for pointing that out (GitHub wouldn't show me enough file context to spot that). I agree -- that code covers this case.

File: clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp
Comment (smanna12) [reply clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp:45]: Done. I have added all test cases above inside clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp.

File: clang/lib/Sema/SemaTemplateInstantiateDecl.cpp
Comment (smanna12) [reply clang/lib/Sema/SemaTemplateInstantiateDecl.cpp:780]: We are handling this inside Sema.h file.

File: clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp
Comment (smanna12) [reply clang/test/CodeGenSYCL/intel-fpga-no-global-work-offset.cpp:16]: Done. Added the test.

File: clang/include/clang/Sema/Sema.h
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h:12975]: We need this diagnostic for the attribute when we pass negative value as template parameter. 

template <int SIZE>
class KernelFunctor {
public:
  // expected-error@+1{{'no_global_work_offset' attribute requires a non-negative integral compile time constant expression}}
  [[intel::no_global_work_offset(SIZE)]] void operator()() {}
};

int main() {
  //expected-note@+1{{in instantiation of template class 'KernelFunctor<-1>' requested here}}
  KernelFunctor<-1>();
  // no error expected
  KernelFunctor<1>();
}

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (smanna12) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: @AaronBallman, thanks for the suggestion. I have updated codegen part.

One issue is here that we can not use the check if (*OptVal) here. This will cause an error.
 error: could not convert ‘OptVal.llvm::Optional<llvm::APSInt>::operator*()’ from ‘llvm::APSInt’ to ‘bool’
     if (*OptVal)
         ^~~~~~~

Comment (AaronBallman) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Oh, that's a think-o on my part, thanks for correcting it. :-)

File: clang/include/clang/Sema/Sema.h
Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h:12975]: I don't see why this is any different than the question of whether to accept values other than 0 or 1. Why are negative numbers a problem? Any nonzero number should be treated as `true`, and zero should be treated as `false`.

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (AaronBallman) [inline clang/lib/CodeGen/CodeGenFunction.cpp:702]: This is testing whether the `Optional<>` has a value or not (which we assert to be true just one line above). This should be `if (ArgVal->getBoolValue())` to look through the `Optional<>` and query the value of the underlying `APSInt`.

File: clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp
Comment (AaronBallman) [inline clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp:12]: I apologize, my suggestion was unclear. I intended for you to instantiate the function template to ensure that the type was checked properly when instantiating from the template definition. You should replace the call here to `func<>()` with:
```
void bar() {
  func<S>();
}
```

Comment (smanna12) [reply clang/test/SemaSYCL/sycl-device-intel-fpga-no-global-work-offset.cpp]: Done

File: clang/lib/CodeGen/CodeGenFunction.cpp
Comment (smanna12) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Done.

File: clang/include/clang/Sema/Sema.h
Comment (smanna12) [reply clang/include/clang/Sema/Sema.h]: Done. I confused myself with the other attributes.  I considered only positive nonzero value as true ( i was wrong). I have removed this diagnostic since no need to check here. Thanks for the clarification.

Comment (AaronBallman) [reply clang/include/clang/Sema/Sema.h]: Thanks!

**Review by AaronBallman — APPROVED** [review]
Thanks for the discussion, the changes LGTM!

**Comment by smanna12** [comment]
> Thanks for the discussion, the changes LGTM!

Thanks @AaronBallman for the reviews and feedback.

**Comment by smanna12** [comment]
Can I get a code-reviews from @Fznamznon / @premanandrao / @elizabethandrews as code-owners? Thanks.

File: clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp
Comment (elizabethandrews) [inline clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp:31]: Please correct me if I've misunderstood this, but any integer constant is now allowed? Does -1 default to 1 here? Can you please add CHECKs for Integer literals as well in this test?

Comment (smanna12) [reply clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp:34]: The attribute takes a Boolean expression. Any nonzero number should be treated as true.
So -1 or 42 is a valid Boolean expression that means true. Zero should be treated as false.

This will make the codes more complicated if we correct the value here (default to 1) 
when creating the semantic attribute. 

Please see the discussion here as a reference.
https://github.com/intel/llvm/pull/2839#discussion_r543398214

After discussion with @AaronBallman, it was decided that if there is a reason, we can make the changes and could correct the value (makes -1 or 42 to default 1).  We will currently leave this now and would not give any diagnostic here.

Comment (elizabethandrews) [reply clang/test/SemaSYCL/intel-fpga-no-global-work-offset.cpp:34]: Oh I see. Thanks for explaining.

**Review by elizabethandrews — APPROVED** [review]

**Comment by smanna12** [comment]
Thank you everyone for the review.

**Pull request merged**
